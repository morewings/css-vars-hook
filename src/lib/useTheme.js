import {useRef, useCallback} from 'react';
import {setCSSVariable, getCSSVariable, createStyleObject, removeCSSVariable} from 'lib/utils';

/**
 * @typedef {Object} UseThemeInterface
 * @property {MutableRefObject<HTMLElement>} ref - React ref. Use as theme container element getter only.
 * @property {function(element: (HTMLElement | null)): void} setRef - Theme container element setter.
 * @property {Object<string, string>} style - Theme object with all property names prefixed `--`.
 * @property {function(string): string} getVariable - Get variable value. getCSSVariable with bound ref.
 * @see src/lib/utils.js
 * @property {function(variableName: string, value: (string|number)): void} setVariable - Set variable value. setCSSVariable with bound ref.
 * @property {function(variableName: string): string} getVariable - Get variable value. getCSSVariable with bound ref.
 * @property {function(variableName: string): void} removeVariable - Remove variable. removeCSSVariable with bound ref.
 */

/** @function
 * @name useTheme
 * @description React hook to apply multiple CSS variables and manipulate them.
 * @param {Object<string, string>} theme - React ref
 * @return {UseThemeInterface}
 */
export const useTheme = theme => {
  const ref = useRef(null);
  const setVariable = setCSSVariable(ref);
  const getVariable = getCSSVariable(ref);
  const removeVariable = removeCSSVariable(ref);
  const getTheme = () => theme;
  const setTheme = nextTheme => {
    Object.keys(nextTheme).forEach(key => {
      const isEqual = getCSSVariable(ref)(key) === theme[key];
      !isEqual && setCSSVariable(ref)(key, nextTheme[key]);
    });
  };
  const style = createStyleObject(theme);
  const setRef = useCallback(
    element => {
      if (ref.current) {
        // Make sure to cleanup any events/references added to the last instance
      }

      if (element) {
        Object.keys(theme).forEach(key => {
          const isEqual = getCSSVariable({current: element})(key) === theme[key];
          !isEqual && setCSSVariable({current: element})(key, theme[key]);
        });
      }

      // Save a reference to the node
      ref.current = element;
    },
    [theme]
  );

  return {ref, setRef, style, setVariable, getVariable, removeVariable, getTheme, setTheme};
};
