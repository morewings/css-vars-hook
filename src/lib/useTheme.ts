import {useRef, useCallback} from 'react';

import {setCSSVariable, getCSSVariable, createStyleObject, removeCSSVariable} from './utils';

/**
 * @deprecated
 * `useLocalTheme` is recommended instead
 * @example
 * const {getTheme, setTheme, LocalRoot} = useLocalTheme()
 */
export const useTheme = (theme: Record<string, string>) => {
    const ref = useRef<HTMLElement>(null);
    const setVariable = setCSSVariable(ref);
    const getVariable = getCSSVariable(ref);
    const removeVariable = removeCSSVariable(ref);
    const getTheme = () => theme;
    const setTheme = (nextTheme: Record<string, string>) => {
        Object.keys(nextTheme).forEach(key => {
            const isEqual = getCSSVariable(ref)(key) === theme[key];
            !isEqual && setCSSVariable(ref)(key, nextTheme[key]);
        });
    };
    const style = createStyleObject(theme);
    const setRef = useCallback(
        (element: HTMLElement) => {
            if (ref.current) {
                // Make sure to cleanup any events/references added to the last instance
            }

            if (element) {
                Object.keys(theme).forEach(key => {
                    const isEqual = getCSSVariable({current: element})(key) === theme[key];
                    !isEqual && setCSSVariable({current: element})(key, theme[key]);
                });
            }

            // Save a reference to the node
            ref.current = element;
        },
        [theme]
    );

    return {ref, setRef, style, setVariable, getVariable, removeVariable, getTheme, setTheme};
};
