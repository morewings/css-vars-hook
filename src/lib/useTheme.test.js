import {renderHook} from '@testing-library/react-hooks';
import {useTheme} from './useTheme';

const theme = {
  foo: 'bar',
};

describe('useTheme', () => {
  let element;

  beforeEach(() => {
    // eslint-disable-next-line fp/no-mutation
    element = document.createElement('div');
  });

  it('renders', () => {
    const {result} = renderHook(() => useTheme(theme));
    expect(result.current).toMatchSnapshot();
  });

  it('`setRef` sets HTMLElement reference to the `ref`', () => {
    const {result} = renderHook(() => useTheme(theme));
    result.current.setRef(element);
    expect(result.current.ref.current).toBe(element);
  });

  it('`style` converts `theme` param to style object', () => {
    const {result} = renderHook(() => useTheme(theme));
    expect(result.current.style).toMatchSnapshot();
  });

  it('`getVariable` gets CSS variable value', () => {
    const {result} = renderHook(() => useTheme(theme));
    const varName = 'foo';
    result.current.setRef(element);
    expect(result.current.getVariable(varName)).toBe(theme[varName]);
  });

  it('`setVariable` sets CSS variable value', () => {
    const {result} = renderHook(() => useTheme(theme));
    const varName = 'foo';
    const varValue = 'baz';
    result.current.setRef(element);
    result.current.setVariable(varName, varValue);
    expect(element.style.getPropertyValue(`--${varName}`)).toBe(varValue);
  });

  it('`removeVariable` removes CSS variable', () => {
    const {result} = renderHook(() => useTheme(theme));
    const varName = 'foo';
    result.current.setRef(element);
    result.current.removeVariable(varName);
    expect(element.style.getPropertyValue(`--${varName}`)).toBe('');
  });
});
