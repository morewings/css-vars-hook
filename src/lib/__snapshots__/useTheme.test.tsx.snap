// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTheme \`style\` converts \`theme\` param to style object 1`] = `
Object {
  "--foo": "bar",
}
`;

exports[`useTheme renders 1`] = `
Object {
  "getTheme": [Function],
  "getVariable": [Function],
  "ref": Object {
    "current": null,
  },
  "removeVariable": [Function],
  "setRef": [Function],
  "setTheme": [Function],
  "setVariable": [Function],
  "style": Object {
    "--foo": "bar",
  },
}
`;
